### ISSUE TITLE
RFC: cross-validation support

### ISSUE THREAD

**Issue by biocompt** [davide-grheco/DataSplits.jl#17]
## Goal

Add cross-validation strategies that complement DataSplits' rational, structure-aware philosophy. Plain random K-fold is already well covered by `MLUtils`/`MLJBase` and is not the package's differentiating value — what is missing is CV that respects groups, time and target distribution, mirroring the same concerns as the existing single-pass strategies.

This issue is an RFC: I would like to settle the design and naming questions below before opening any implementation PR, so the actual code review can focus on correctness and not on shape.

## Why now

After #16 (`WithValidation`), the only result type in `core.jl` without a producer is `CrossValidationSplit`. Closing that gap rounds out the post-refactor API and unlocks the typical model-selection workflows users come to DataSplits for.

## Proposed v1 scope

| Strategy | Slot | Notes |
|---|---|---|
| `GroupKFold(k)` | `:groups` | No group spans folds; high value, natural extension of `GroupShuffleSplit`. |
| `TimeSeriesCV(k; gap=0)` | `:time` | Expanding (or rolling) window with optional gap; mirrors `TimeSplit`. |
| `StratifiedKFold(k; ...)` | `:target` | Classes for classification, binned for regression. |
| `LeaveOneGroupOut()` | `:groups` | Trivial reduction of `GroupKFold` once that exists. |

Each returns `CrossValidationSplit{TrainTestSplit}`. No changes to `partition` / `splitdata` / `splitview` / `consumes` / `fallback_from_data` are required — the `CrossValidationSplit` dispatch is already in place.

## Out of scope (v1)

- Plain random `KFold` — delegate to `MLUtils`.
- Repeated / nested / Monte-Carlo CV.
- CV variants of distance-based strategies (Kennard–Stone over K folds, OptiSim over K folds, …). High-value research direction and arguably the most differentiating thing DataSplits could offer, but the contract benefits from being settled on simpler strategies first.

## Open architectural questions

These are the points I want feedback on before any code is written.

### 1. Fold result type

Options:
- **A**: CV strategies return `CrossValidationSplit{TrainTestSplit}` only. Users that want a held-out validation cohort per fold compose with `WithValidation` themselves.
- **B**: Allow `CrossValidationSplit{TrainValTestSplit}` natively, e.g. via a `WithValidation` over a CV strategy, or via a `val_frac` parameter on each CV strategy.

I lean towards **A** for v1 — it keeps each strategy single-purpose and reuses #16 cleanly. **B** is reachable later by making `WithValidation` also accept a CV strategy as outer (currently it errors); that is a natural follow-up but a separate PR.

### 2. `StratifiedKFold` over continuous targets

How should binning work for regression?

- Explicit `bins` keyword (user passes the number of bins, or the bin edges).
- Automatic binning with a sensible default (e.g. `nbins = ceil(sqrt(N))` or quantile-based).
- Reject continuous targets (classification only) and provide a separate `QuantileKFold` for regression.

I would default to **explicit `bins::Int = 10`** and document the choice; auto-magic feels out of character with the rest of the package.

### 3. `TimeSeriesCV` window semantics

- Expanding window vs. rolling window — which is the default?
- `gap` semantics: integer number of observations skipped between train and test, or a duration that depends on `eltype(time)`?
- Alignment when timestamps are non-unique (groups of same-day observations) — same rule as `TimeSplit`?

I would propose **expanding window default**, **`gap::Int` in observations** for simplicity, and **same grouping rule as `TimeSplit`**. Open to alternatives.

### 4. Naming convention

Tied to the question raised in #16. Every strategy in the package today ends in `Split` (`KennardStoneSplit`, `GroupShuffleSplit`, …). For CV that suffix arguably makes less sense. Two options:

- `KFold`, `GroupKFold`, `TimeSeriesCV`, `StratifiedKFold`, `LeaveOneGroupOut` — short, conventional, matches `MLUtils`/scikit-learn naming.
- `KFoldSplit`, `GroupKFoldSplit`, … — keeps the suffix uniform across the package.

I lean towards the short form (and would consider renaming `WithValidation` accordingly to keep the package internally consistent), but happy to go either way.

### 5. PR granularity

- One bundled `feat: cross-validation strategies` PR with all four.
- Four small PRs in sequence: `GroupKFold` first as the foundation, then the others in any order.

I prefer **four small PRs** — easier review, each lands when ready, and `GroupKFold` is the smallest end-to-end example to lock in the patterns (struct shape, `_partition` signature for CV, test layout, doc page) before the rest follow.

## Trait surface (no changes expected)

Each CV strategy declares its consumed slots exactly like existing single-pass strategies:

```julia
consumes(::GroupKFold) = (:groups,)
fallback_from_data(::GroupKFold) = (:groups,)

function _partition(data, alg::GroupKFold; groups, rng, kwargs...)
    folds = ... # Vector{TrainTestSplit}
    return CrossValidationSplit(folds)
end
```

`partition`, `splitdata`, `splitview` already dispatch on `CrossValidationSplit`. Nothing in `core.jl` should need to change.

## Asks

@davide-grheco — when you have a moment, would you weigh in on the five open questions? Anything I have missed is also welcome. Once those are settled I will open the first implementation PR (`GroupKFold`) and we can iterate from there.

**Comment by davide-grheco** [comment]
I guess a nice interface for the user could be:

```
for split in cvs
    train, test = splitview(data, split)
end
```

and 
```
length(cvs)
first(cvs)
cvs[1]
collect(cvs)
```

It would also be nice to provide something like this:

```
for (train, test) in splitview(data, cvs)
end
```

and, just in case:

```
for (train_idx, test_idx) in indices(cvs)
end
```

I guess most of this can be achieved by just returning a Vector and implementing methods for `indices`, `splitview` and other functions to work on vectors.

**Comment by biocompt** [comment]
@davide-grheco — taking a position on the five RFC questions and a sixth that arose during prototyping. Opening PR #N right after this comment so the architectural pieces can be reviewed in code; happy to revise either way once you weigh in here.

**1. Fold result type** → **Option A** for v1 (`CrossValidationSplit{TrainTestSplit}` only). With the 3-cohort `partition(data, test_alg, val_alg; train, validation, test)` already merged on `api`, train/val/test composition has a clean home outside CV; mixing them now would just pre-empt a follow-up.

**2. `StratifiedKFold` continuous targets** → **explicit `bins::Int = 10`**, no auto-magic. Float `eltype` triggers quantile binning; everything else (Int, Bool, BitVector, Symbol, String) is treated as discrete classes. Documented that targets with many duplicate values can collapse quantile edges.

**3. `TimeSeriesCV` semantics** → **expanding window default**, **rolling via `rolling=true`**, **`gap::Int` in observations**, atomicity by timestamp identical to `TimeSplit` (chunk boundaries fall between distinct time values). One subtle point: `gap` is positional, so when it falls inside a block of equal timestamps it can split that block on the train side. The contract guarantees no leakage (no row appears in both train and test of the same fold), but the train side is no longer block-aligned — documented explicitly.

**4. Naming** → **short form**: `GroupKFold`, `LeaveOneGroupOut`, `TimeSeriesCV`, `StratifiedKFold`. Aligns with sklearn/MLUtils. Not adding a `Split` suffix to CV strategies.

**5. PR granularity** → **4 small PRs in sequence** as you suggested. PR 1 (`GroupKFold` + foundation) opening now; the other three are ready in local branches and will go up after PR 1 lands.

**6. New decision (not in the RFC)** → introduced an **`AbstractCVStrategy <: AbstractSplitStrategy`** marker type and a third `partition(data, alg::AbstractCVStrategy; …)` method without `train`/`test` kwargs. The existing 2-cohort `partition` requires `train`/`test` (after PR #20), and CV strategies don't have those — fold sizes are determined by `k`. Adding a parallel `partition` method dispatched on a marker subtype mirrors exactly the pattern you already use for the 2-cohort and 3-cohort forms; alternatives (making `train`/`test` optional and runtime-branching, or trait-based dispatch) felt strictly worse in error reporting and contract clarity. PR 1 contains the type and the new method as the foundation; if you'd rather take a different path here, this is the spot to flag it.

**Bonus** (matches your iteration comment from this thread): added `Base.getindex` / `firstindex` / `lastindex` / `eltype` on `CrossValidationSplit`, so `cvs[1]`, `cvs[end]`, `cvs[range]`, `first(cvs)`, `last(cvs)`, `collect(cvs)` and `for (train, test) in splitview(cvs, X)` all work.

PR coming next.

**Comment by biocompt** [comment]
@davide-grheco — closing summary against the original RFC. All five open questions are now settled in shipped code on `api`, and the v1 scope plus the broader scikit-learn parity push are in place. I think this issue can be closed.

## v1 scope (this RFC)

| Strategy           | Slot     | Landed in |
|--------------------|----------|-----------|
| `GroupKFold`       | `:groups`| `api`     |
| `TimeSeriesSplit`  | `:time`  | #34       |
| `StratifiedKFold`  | `:target`| `api`     |
| `LeaveOneGroupOut` | `:groups`| `api`     |

## Architectural decisions, as shipped

1. **Fold result type** → Option A. CV strategies return `CrossValidationSplit{TrainTestSplit}`. Train/val/test composition is handled by the 3-cohort `partition(data, test_alg, val_alg; train, validation, test)` that was already on `api`.
2. **`StratifiedKFold` on continuous targets** → explicit `bins::Int = 10`, quantile binning for `Float` `eltype`; everything else treated as discrete classes.
3. **`TimeSeriesSplit` semantics** → expanding window by default, rolling via `max_train_size`, `gap::Int` in observations, atomicity by timestamp (chunk boundaries fall between distinct time values, matching `TimeSplit`).
4. **Naming** → short form (`GroupKFold`, `LeaveOneGroupOut`, `TimeSeriesSplit`, `StratifiedKFold`) — no `Split` suffix for CV strategies.
5. **PR granularity** → small focused PRs (one strategy per PR), as you preferred.
6. **`AbstractCVStrategy` marker type + dedicated `partition` method without `train`/`test` kwargs** — added as the foundation. Fold sizes determined by the strategy's own parameters (`k`, `n_splits`, …).
7. **Iteration protocol on `CrossValidationSplit`** (your follow-up comment) — `cvs[i]`, `first(cvs)`, `last(cvs)`, `collect(cvs)`, `for (train, test) in splitview(cvs, X)` all work.

## Beyond the original v1

The CV surface has grown well past the four strategies originally proposed:

- `KFold`, `ShuffleSplit`, `StratifiedShuffleSplit`, `GroupShuffleSplitCV` (#27 / PR #50), `StratifiedGroupKFold`, `RepeatedKFold`, `RepeatedStratifiedKFold`, `LeavePOut`, `LeavePGroupsOut`, `PredefinedSplit`, `BootstrapSplit`, `BlockedCV`.
- scikit-learn parity is tracked in #27; the remaining checkbox there is closed by PR #50.
- Additional CV methods not in scikit-learn (Purged K-Fold, Nested CV, Monte Carlo CV, …) are tracked in #28.

If you agree, this issue can be closed and #27 / #28 carry the remaining work.
